common/src/com/thoughtworks/go/domain/materials/svn/SvnMaterialUpdater.java

```diff
@@ -0,0 +1,85 @@
+package com.thoughtworks.go.domain.materials.svn;
+
+import com.thoughtworks.go.config.materials.svn.SvnMaterial;
+import com.thoughtworks.go.domain.BuildCommand;
+import com.thoughtworks.go.domain.materials.Revision;
+import com.thoughtworks.go.domain.materials.RevisionContext;
+import com.thoughtworks.go.util.StringUtil;
+import org.apache.commons.lang.StringUtils;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.thoughtworks.go.domain.BuildCommand.*;
+
+public class SvnMaterialUpdater {
+    private SvnMaterial material;
+
+    public SvnMaterialUpdater(SvnMaterial material) {
+        this.material = material;
+    }
+
+    public BuildCommand updateTo(String baseDir, RevisionContext revisionContext) {
+        Revision revision = revisionContext.getLatestRevision();
+        String workingDir = material.workingdir(new File(baseDir)).getPath();
+
+        BuildCommand shouldCheckout = or(
+                test("-nd", workingDir),
+                test("-nd", new File(workingDir, ".svn").getPath()),
+                isRepositoryChanged(workingDir));
+
+        return compose(
+                echoWithPrefix("Start updating %s at revision %s from %s", material.updatingTarget(), revision.getRevision(), material.getUrl()),
+                cond(shouldCheckout,
+                        freshCheckout(workingDir, revision),
+                        cleanupAndUpdate(workingDir, revision)),
+                echoWithPrefix("Done.\n")
+        );
+
+    }
+
+    private BuildCommand cleanupAndUpdate(String workingDir, Revision revision) {
+        return compose(
+                svn("cleanup", workingDir),
+                svn("revert", "--recursive", workingDir),
+                svnWithAuth("update", "--non-interactive", "-r", revision.getRevision(), workingDir));
+    }
+
+    private BuildCommand isRepositoryChanged(String workingDir) {
+        return test("-ncontains",
+                "URL: " + StringUtil.removeTrailingSlash(material.getUrl()),
```

Comment:
I think there's a newline missing here. It fails this test:

```
@Test
public void shouldCheckoutIfSvnRepositoryChangesToAParentRepo() throws IOException {
    String initialRepo = testRepo.projectRepositoryUrl();
    runBuild(buildCommand("3", initialRepo), JobResult.Passed);
    assertThat(console.output(), containsString("Checked out revision 3"));

    String changedRepo = FileUtil.toFileURI(testRepo.projectRepositoryRoot());
    runBuild(buildCommand("4", changedRepo), JobResult.Passed);
    assertThat(console.output(), containsString("Checked out revision 4"));

    assertThat(getWorkingDirInfo(workingCopy).getUrl() + "/", is(changedRepo));
}
```

If that is right, you should probably take a look at removing that `+ "/"` in the test.